tempesta_fw/http_parser.c

```diff
@@ -1370,6 +1393,332 @@ __parse_content_type(TfwHttpMsg *hm, unsigned char *data, size_t len)
 	return r;
 }
 
+#define GROW_STR(str, data, sz)						\
+do {									\
+	TfwStr *last_chunk = TFW_STR_LAST(str);				\
+	if ((str)->len > 0 &&						\
+	    (char *)(str)->ptr + (str)->len == (char *)(data))		\
+	{								\
+		if (last_chunk != (str))				\
+			last_chunk->len += (sz);			\
+		(str)->len += (sz);					\
+	} else {							\
+		__msg_field_fixup_pos(str, data, sz);			\
+	}								\
+} while (0)
+
+/*
+ * Resolves all escaped characters, removes heading and trailing double quotes.
+ * Calling site should ensure dst->ptr have enough writable memory. Resulting
+ * string is at most src->len bytes long.
+ */
+static void
+unescape_multipart_boundary(TfwStr * __restrict dst, TfwStr * __restrict src)
+{
+	TfwStr *c, *c_end;
+	bool esc = false;
+	char *pos;
+	char *last;
+
+	/* There are at least two double quotes. */
+	BUG_ON(src->len < 2);
+
+	if (TFW_STR_PLAIN(src)) {
+		c = src;
+		c_end = src + 1;
+	} else {
+		c = src->ptr;
+		c_end = c + TFW_STR_CHUNKN(src);
+	}
+
+	BUG_ON(c->len == 0);
+	pos = (char *)c->ptr + 1;
+	last = (char *)c->ptr + c->len;
+	dst->len = 0;
+	while (1) {
+		if (pos == last) {
+			c += 1;
+			if (c == c_end)
+				break;
+			pos = c->ptr;
+			last = pos + c->len;
+		}
+
+		if (esc) {
+			((char *)dst->ptr)[dst->len++] = *pos++;
+			esc = false;
+			continue;
+		}
+
+		if (unlikely(*pos == '"'))
+			return;
+
+		if (*pos == '\\') {
+			esc = true;
+			pos += 1;
+			continue;
+		}
+
+		((char *)dst->ptr)[dst->len++] = *pos++;
+	}
+}
+
+static int
+__req_parse_content_type(TfwHttpMsg *hm, unsigned char *data, size_t len)
```

Comment:
The whole function doesn't follow approach used all across http_parser. As a result of this function we got the following:
- `req->h_tbl[TFW_HTTP_HDR_CONTENT_TYPE]` -> TfwStr with no special limiting by chunks
- newly allocated `req->multipart_boundary_raw` which points to boundary value with its own splitting by chunks
- newly allocated `req->multipart_boundary`  which points to parts of boundary value with its own splitting by chunks
- copy operation from `req->multipart_boundary_raw` to `req->multipart_boundary` which is done char-by-char. And actually the same work is done twice: first we scan for escapes in fsm, and the second time in `unescape_multipart_boundary()`.

To follow the approach used in the parser module, the function should:
1. Call `__TFW_HTTP_PARSE_SPECHDR_VAL(......, 0)` macro. This macro doesn't fill parsed header, instead parsed header is filled manually chunck-by-chunk, each chunk can have very specific flags. In this case there is no need to allocate additional space for `req->multipart_boundary_raw` or make full symbols scans in `unescape_multipart_boundary()`. See `__req_parse_cookie()` as an example.

2. Parse the header until start of boundary value. Save all parsed chunks into `parser->hdr` (see `__FSM_I_MOVE_fixup()` macro)

3. Next chunk is start of boundary value, set `req->multipart_boundary_raw->ptr` to `TFW_STR_CURR(parser->hdr)`

4. If the currently parsed symbol should appear in unescaped `req->multipart_boundary`, add it to current chunk and set `TFW_STR_VALUE` flag for it. 

5. If current character is escape character and should not appear in `req->multipart_boundary`, put it in it's own chunk without `TFW_STR_VALUE` flag.

6. Loop 4-5 until boundary value end.

7. Once boundary value ended, collect `req->multipart_boundary` and `req->multipart_boundary_raw`. Number of chunks for `req->multipart_boundary_raw` can be easily calculated: `n = TFW_STR_CURR(parser->hdr) -req->multipart_boundary_raw->ptr. And then run through `req->multipart_boundary_raw->ptr` and copy TfwStr headers to  `req->multipart_boundary` if `TFW_STR_VALUE` flag set. This copy operation is faster, since it scans by chanks, not by symbols.

8. Save the rest of data into `parser->hdr`

Reply:
Thanks for the detailed instructions!
I've changed the patch to follow them, but rather than using a pointer from `TFW_STR_CURR(parser->hdr)`, used a current chunk number. It turned out, that during `parser->hdr` growth, `parser->hdr.ptr` is reallocated multiple times. At some point, allocator may change `parser->hdr.ptr`.